pkg/auth/auth.go

```diff
@@ -0,0 +1,250 @@
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/adal"
+	"github.com/dimchansky/utfbom"
+	"github.com/virtual-kubelet/virtual-kubelet/log"
+)
+
+type CloudEnvironmentName string
+
+const (
+	AzurePublicCloud       CloudEnvironmentName = "AzurePublicCloud"
+	AzureUSGovernmentCloud CloudEnvironmentName = "AzureUSGovernment"
+	AzureChinaCloud        CloudEnvironmentName = "AzureChina"
+)
+
+type Config struct {
+	AKSCredential *aksCredential
+	AuthConfig    *Authentication
+	Cloud         cloud.Configuration
+	Authorizer    autorest.Authorizer
+}
+
+// getAuthorizer return autorest authorizer.
+func (c *Config) getAuthorizer(resource string) (autorest.Authorizer, error) {
+	var auth autorest.Authorizer
+	var err error
+
+	var token *adal.ServicePrincipalToken
+	isUserIdentity := len(c.AuthConfig.ClientID) == 0
+
+	if isUserIdentity {
+		token, err = adal.NewServicePrincipalTokenFromManagedIdentity(
+			resource, &adal.ManagedIdentityOptions{ClientID: c.AuthConfig.UserIdentityClientId})
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		oauthConfig, err := adal.NewOAuthConfig(
+			c.Cloud.ActiveDirectoryAuthorityHost, c.AuthConfig.TenantID)
+		if err != nil {
+			return nil, err
+		}
+		token, err = adal.NewServicePrincipalToken(
+			*oauthConfig, c.AuthConfig.ClientID, c.AuthConfig.ClientSecret, resource)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	auth = autorest.NewBearerAuthorizer(token)
+	return auth, err
+}
+
+// SetAuthConfig sets the configuration needed for Authentication.
+func (c *Config) SetAuthConfig() error {
+	var err error
+	c.Cloud = cloud.AzurePublic
+
+	if authFilepath := os.Getenv("AZURE_AUTH_LOCATION"); authFilepath != "" {
+		auth := &Authentication{}
+		err = auth.newAuthenticationFromFile(authFilepath)
+		if err != nil {
+			return err
+		}
+		c.AuthConfig = auth
+	}
+
+	if aksCredFilepath := os.Getenv("AKS_CREDENTIAL_LOCATION"); aksCredFilepath != "" {
+		c.AKSCredential, err = newAKSCredential(aksCredFilepath)
+		if err != nil {
+			return err
+		}
+
+		var clientId string
+		if !strings.EqualFold(c.AKSCredential.ClientID, "msi") {
+			clientId = c.AKSCredential.ClientID
+		}
+
+		//Set Azure cloud environment
+		c.Cloud = getCloudConfiguration(c.AKSCredential.Cloud)
+		c.AuthConfig = NewAuthentication(
+			clientId,
+			c.AKSCredential.ClientSecret,
+			c.AKSCredential.SubscriptionID,
+			c.AKSCredential.TenantID,
+			c.AKSCredential.UserAssignedIdentityID)
+
+		if clientID := os.Getenv("AZURE_CLIENT_ID"); clientID != "" {
+			c.AuthConfig.ClientID = clientID
+		}
+
+		if clientSecret := os.Getenv("AZURE_CLIENT_SECRET"); clientSecret != "" {
+			c.AuthConfig.ClientSecret = clientSecret
+		}
+
+		if userIdentityClientId := os.Getenv("VIRTUALNODE_USER_IDENTITY_CLIENTID"); userIdentityClientId != "" {
+			c.AuthConfig.UserIdentityClientId = userIdentityClientId
+		}
+
+		isUserIdentity := len(c.AuthConfig.ClientID) == 0
+
+		if isUserIdentity {
```

Comment:
You can further remove "isUserIdentity" variable and add a comment instead. Feel free to do this in next change.